scrapy_zyte_api/providers.py

```diff
@@ -112,10 +125,35 @@ async def __call__(  # noqa: C901
                     options["extractFrom"] = extract_from.value
                     break
 
+        http_response_needed = (
+            AnyResponse in to_provide
+            and BrowserResponse not in to_provide
+            and BrowserHtml not in to_provide
+            and not http_response
+        )
+
+        extract_from = None  # type: ignore[assignment]
         for item_type, kw in item_keywords.items():
             options_name = f"{kw}Options"
             if item_type not in to_provide_stripped and options_name in zyte_api_meta:
                 del zyte_api_meta[options_name]
+            elif options_name in zyte_api_meta:
+                extract_from = zyte_api_meta[options_name]["extractFrom"]
+            elif item_type in to_provide_stripped and http_response_needed:
+                zyte_api_meta[options_name] = {"extractFrom": "httpResponseBody"}
```

Comment:
Could `"ZYTE_API_PROVIDER_PARAMS": {"productOptions": {}}` trigger a `KeyError` here?

Reply:
I don't think so, since before this line, the following is executed for each item type. 

```python
options = zyte_api_meta.setdefault(f"{kw}Options", {})
options["extractFrom"] = extract_from.value
```

Though I agree it'd be better to have `extract_from = zyte_api_meta[options_name].get("extractFrom")` instead, in case some future refactor changes this assumption. I'll update this. 👍